File: android/app/src/main/java/com/reactnativemaestroexample/NetworkModule.java
package com.reactnativemaestroexample;
import android.content.Context;
import android.net.ConnectivityManager;
import android.net.Network;
import android.net.NetworkCapabilities;
import android.net.NetworkRequest;
import android.os.Build;
import android.util.Log;

import androidx.annotation.NonNull;

import com.facebook.react.bridge.LifecycleEventListener;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;

public class NetworkModule extends ReactContextBaseJavaModule implements LifecycleEventListener {
    private static int selectedNetworkType = 1; // 1 = wifi, 0 = mobile data
    private ReactApplicationContext appContext;
    NetworkModule(ReactApplicationContext context) {
        super(context);
        appContext = context;
        context.addLifecycleEventListener(this);
    }

    @Override
    public String getName() {
        return "NetworkModule";
    }

    @ReactMethod
    public void changeNetworkToMobileData() {
        Log.d("NetworkModule", "Changing Network to Mobile Data");
        selectedNetworkType = 0;
        forceConnection(appContext);
    }

    @ReactMethod
    public void changeNetworkToWifi() {
        Log.d("NetworkModule", "Changing Network both mobile and wifi");
        selectedNetworkType = 1;
        forceConnection(appContext);
    }

    public static void forceConnection(ReactApplicationContext context) {
        final ConnectivityManager connectivityManager = (ConnectivityManager) context.getSystemService(Context.CONNECTIVITY_SERVICE);
        if(Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            NetworkRequest.Builder request = new NetworkRequest.Builder();
            Log.d("NetworkModule", "Request Transport Cellular");
            if(selectedNetworkType == 0) {
                request.addTransportType(NetworkCapabilities.TRANSPORT_CELLULAR);
                request.addCapability(NetworkCapabilities.NET_CAPABILITY_INTERNET);
            }

            connectivityManager.requestNetwork(request.build(), new ConnectivityManager.NetworkCallback() {
                @Override
                public void onAvailable(Network network) {
                    super.onAvailable(network);
                    if(Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
                        if(selectedNetworkType == 0) {
                            Log.d("NetworkModule", "Binding app to cellular");
                            connectivityManager.bindProcessToNetwork(network);
                        } else {
                            Log.d("NetworkModule", "Binding app to default");
                            connectivityManager.bindProcessToNetwork(null);
                        }
                    }
                }

                @Override
                public void onUnavailable() {
                    Log.d("NetworkModule", "Unavailable");
                }

                @Override
                public void onCapabilitiesChanged(@NonNull Network network, @NonNull NetworkCapabilities networkCapabilities) {
                    super.onCapabilitiesChanged(network, networkCapabilities);
                    final boolean unmetered = networkCapabilities.hasCapability(NetworkCapabilities.NET_CAPABILITY_NOT_METERED);
                    Log.d(
                            "NetworkModule",
                            "NetworkStatusChanged: \nUnmetered: " + unmetered + "\nInternetAvailable: " + networkCapabilities.hasCapability(NetworkCapabilities.NET_CAPABILITY_INTERNET) + "\nNetViaMobileInternet: " + networkCapabilities.hasTransport(NetworkCapabilities.TRANSPORT_CELLULAR) + "\nNetViaWIFI: " + networkCapabilities.hasTransport(NetworkCapabilities.TRANSPORT_WIFI)
                    );
                }

            });
        }
    }

    @Override
    public void onHostResume() {
        Log.d("NetworkModule", "onHostResume");
        forceConnection(appContext);
    }

    @Override
    public void onHostPause() {
        Log.d("NetworkModule", "onHostPause");
    }

    @Override
    public void onHostDestroy() {
        Log.d("NetworkModule", "onHostDestroy");
    }
}

File: android/app/src/main/java/com/reactnativemaestroexample/NetworkModule.kt
package com.reactnativemaestroexample

import android.net.ConnectivityManager
import android.net.LinkProperties
import android.net.Network
import android.net.NetworkCapabilities
import android.net.NetworkRequest
import android.util.Log
import com.facebook.react.bridge.Promise
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactContextBaseJavaModule
import com.facebook.react.bridge.ReactMethod

class NetworkModule(reactContext: ReactApplicationContext): ReactContextBaseJavaModule(reactContext) {
    private val appContext = reactContext;
    private val connectivityManager = appContext.getSystemService(ConnectivityManager::class.java);
    private val TAG = "NetworkModule";

    override fun getName() = "NetworkModule";

    @ReactMethod fun changeNetworkToCellular(promise: Promise) {
        // TODO: Before binding check if mobile data is on
        bindAppToCellularTransportNetwork(promise);
    }

    @ReactMethod fun changeNetworkToWifi() {
        bindAppToDefaultNetwork();
    }

    private fun bindAppToCellularTransportNetwork(promise: Promise) {
        val cellularNetworkRequest = NetworkRequest.Builder()
            .addTransportType(NetworkCapabilities.TRANSPORT_CELLULAR)
            .addCapability(NetworkCapabilities.NET_CAPABILITY_INTERNET)
            .build();

        connectivityManager.requestNetwork(cellularNetworkRequest, object : ConnectivityManager.NetworkCallback() {
            override fun onAvailable(network : Network) {
                printNetwork("Available Network: ", network)
                connectivityManager.bindProcessToNetwork(network)
                promise.resolve("Switched to cellular");
            }

            override fun onLosing(network: Network, maxMsToLive: Int) {
                printNetwork("Losing Network: ", network)
            }

            override fun onLost(network : Network) {
                printNetwork("Lost Network: ", network)
            }

            override fun onUnavailable() {
                Log.d(TAG, "Unavailable Network")
                promise.reject("SwitchToCellular", "Unable to switch, probably data is off")
            }

            override fun onCapabilitiesChanged(network : Network, networkCapabilities : NetworkCapabilities) {
                printNetwork("Capabilities changed : ", network)
            }

            override fun onLinkPropertiesChanged(network : Network, linkProperties : LinkProperties) {
                Log.d(TAG, "Link properties changed: $linkProperties")
            }
        }, 1000)
    }

    private fun bindAppToDefaultNetwork() {
        Log.d(TAG, "Listen to wifi")
        connectivityManager.bindProcessToNetwork(null);
    }

    private fun printNetwork(prefix: String, network: Network) {
        val currActiveNetworkCap = connectivityManager.getNetworkCapabilities(network);
        if(currActiveNetworkCap !== null) {
            val isCellular = currActiveNetworkCap.hasTransport(NetworkCapabilities.TRANSPORT_CELLULAR);
            val isWifi = currActiveNetworkCap.hasTransport(NetworkCapabilities.TRANSPORT_WIFI);
            Log.d(TAG, "$prefix\nisCellular: $isCellular\nisWifi: $isWifi");
        }
    }
}